tests/test_representations.py

```diff
@@ -0,0 +1,33 @@
+import flask_restplus.representations as rep
+
+from json import dumps, loads
+from ujson import dumps as udumps, loads as uloads
+
+from flask import current_app
+
+
+def test_representations(api):
+    payload = {
+        'id': 1,
+        'name': 'toto',
+        'address': 'test',
+    }
+    r = rep.output_json(payload, 200)
+    assert loads(r.get_data(True)) == loads(dumps(payload))
+    # now reset serializer
+    rep.serializer = None
+    # then enforce a custom serializer
+    current_app.config['RESTPLUS_JSON_SERIALIZER'] = 'ujson'
+    r2 = rep.output_json(payload, 200)
+    assert uloads(r2.get_data(True)) == uloads(udumps(payload))
```

Comment:
This is testing several different things within a single test case - how do you feel about splitting it up into several test cases so that it's clearer to see the individual assertions? Something like:
```python
def test_representations_serialization_output_correct():
    # original test comparing get_data and dumps outputs

def test_config_custom_serializer_is_module():
    # current_app.config['RESTPLUS_JSON_SERIALIZER'] = 'ujson'
    # test .dumps method is used 

def test_config_custom_serializer_is_function():
    # current_app.config['RESTPLUS_JSON_SERIALIZER'] = 'my_custom_json.module.dump_method'
    # test correct function imported

def test_config_custom_serializer_fallback():
    # test fallback to json.dumps
```
This makes the intended behaviour clearer and is easier to see from a test failure where the issue is :smile:

Reply:
Thanks!

I have spit the test.
I'm not quite sure about the `warnings.warn` yet though. I think we should define a *communication with upper apps* policy first so people know how to deal with those.

Reply:
Great thank you! Sorry, I'm not sure what you mean by
> communication with upper apps policy

Could you please clarify? Did you mean define a way in which Flask-RESTPlus should interact with other applications?

Using `warnings/warn` is a common practice for Flask extensions that wish to communicate misconfiguration/possible error to the user. For example:
- [Flask-Login](https://github.com/maxcountryman/flask-login/blob/7e451269fbb7d58ad2530452a2d3420e0766ce3a/flask_login/login_manager.py#L100)
- [Flask-Debugtoolbar](https://github.com/mgood/flask-debugtoolbar/blob/d852042ccb68d0507e0ef8d9e0d53287ca766f3b/flask_debugtoolbar/__init__.py#L214)
- [Flask-Limiter](https://github.com/alisaifee/flask-limiter/blob/c0ae4d944bf233dccd68207c4de9e2c242cb055f/flask_limiter/extension.py#L112)

Or using `logger.warning` is also a common approach.

Reply:
> Could you please clarify? Did you mean define a way in which Flask-RESTPlus should interact with other applications?

I mean we should document what kind of *communication* people using FLask-RESTPlus should expect.
I know `warnings.warn` is a common practice, but we should document this.
The reason is the default filter might hide some messages hence this doesn't provide any help to the users:

```python
import warnings

warnings.warn('You should not use this', DeprecationWarning)
warnings.warn('Hey, something went wrong', UserWarning)
```

Result in:

```
$ python /tmp/warn.py
/tmp/warn.py:4: UserWarning: Hey, something went wrong
  warnings.warn('Hey, something went wrong', UserWarning)
```

So by default, unless people know they should expect a `DeprecationWarning` they won't notice it.

Reply:
Ah yes I understand - thank you for clarifying :smile: Yes I agree that this should definitely be documented. How about changing this:
https://github.com/noirbizarre/flask-restplus/pull/637/files#diff-dce4048f760a2254db20f3a78d7fcfa1R362 to say something like:

> Flask-RESTPlus will always fallback to the default ``json.dumps``  *serializer* if it cannot manage to import the one you configured. This is indicated by a ``warning`` during initialisation.

And then include an example of the output when starting the application, for example using `logging`:

```
* Serving Flask app "t" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: on
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
 * Restarting with stat
unable to import RESTPLUS_JSON_SERIALIZER=broken_seralizer.dumps, falling back to json.dumps
 * Debugger is active!
 * Debugger PIN: 627-628-298
```

Or using `warnings`:
```
 * Serving Flask app "t" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: on
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
 * Restarting with stat
/home/ben/Documents/flask-restplus/flask_restplus/api.py:213: UserWarning: unable to import RESTPLUS_JSON_SERIALIZER=broken_seralizer.dumps, falling back to json.dumps
  warnings.warn("unable to import RESTPLUS_JSON_SERIALIZER=broken_seralizer.dumps, falling back to json.dumps")
 * Debugger is active!
 * Debugger PIN: 627-628-298
```

Reply:
The only problem I see with this is the *code path* in which this warning will be raised is not deterministic (it will happen only after **some** *route* returns and if it needs marshalling).

Maybe the solution is to load the serializer within the `Api` constructor.

This way, people know when/where to catch the `warnings`.

Reply:
I have added the `warnings.warn` as suggested.
This should happen early in the initialization of the Api unless you pass it a Blueprint instead of a Flask app (in which case the warning will be raised when calling a route that needs serialization.
The loaded serializer is *cached* within the `app.config`. I don't think this is the best place to do so, but I didn't find any other...